lighthouse-core/audits/byte-efficiency/bundle-duplication.js

```diff
@@ -0,0 +1,224 @@
+   * @param {Array<LH.Artifacts.NetworkRequest>} networkRecords
+   * @param {LH.Audit.Context} context
+   * @return {Promise<ByteEfficiencyAudit.ByteEfficiencyProduct>}
+   */
+  static async audit_(artifacts, networkRecords, context) {
+    const ignoreThresholdInBytes =
+      context.options && context.options.ignoreThresholdInBytes || IGNORE_THRESHOLD_IN_BYTES;
+
+    const duplication =
+      await BundleDuplication._getDuplicationGroupedByNodeModules(artifacts, context);
+
+    /**
+     * @typedef ItemSubrows
+     * @property {string[]} urls
+     * @property {number[]} sourceBytes
+     */
+
+    /**
+     * @typedef {LH.Audit.ByteEfficiencyItem & ItemSubrows} Item
+     */
+
+    /** @type {Item[]} */
+    const items = [];
+
+    /** @type {Map<string, number>} */
+    const wastedBytesByUrl = new Map();
+    for (const [source, sourceDatas] of duplication.entries()) {
+      // One copy of this module is treated as the canonical version - the rest will have
+      // non-zero `wastedBytes`. In the case of all copies being the same version, all sizes are
+      // equal and the selection doesn't matter. When the copies are different versions, it does
+      // matter. Ideally the newest version would be the canonical copy, but version information
+      // is not present. Instead, size is used as a heuristic for latest version. This makes the
+      // audit conserative in its estimation.
+
+      const urls = [];
+      const bytesValues = [];
+      let wastedBytesTotal = 0;
+      for (let i = 0; i < sourceDatas.length; i++) {
+        const sourceData = sourceDatas[i];
+        const url = sourceData.scriptUrl;
+        urls.push(url);
+        bytesValues.push(sourceData.size);
+        if (i === 0) continue;
+        wastedBytesTotal += sourceData.size;
+        wastedBytesByUrl.set(url, (wastedBytesByUrl.get(url) || 0) + sourceData.size);
+      }
+
+      items.push({
+        source,
+        wastedBytes: wastedBytesTotal,
+        // Not needed, but keeps typescript happy.
+        url: '',
+        // Not needed, but keeps typescript happy.
+        totalBytes: 0,
+        urls,
+        sourceBytes: bytesValues,
+      });
+    }
+
+    /** @type {Item} */
+    const otherItem = {
+      source: 'Other',
+      wastedBytes: 0,
+      url: '',
+      totalBytes: 0,
+      urls: [],
+      sourceBytes: [],
+    };
+    for (const item of items.filter(item => item.wastedBytes <= ignoreThresholdInBytes)) {
+      otherItem.wastedBytes += item.wastedBytes;
+      for (let i = 0; i < item.urls.length; i++) {
+        const url = item.urls[i];
+        if (!otherItem.urls.includes(url)) {
+          otherItem.urls.push(url);
+        }
+      }
+      items.splice(items.indexOf(item), 1);
+    }
+    if (otherItem.wastedBytes > ignoreThresholdInBytes) {
+      items.push(otherItem);
+    }
+
+    // Convert bytes to transfer size estimation.
+    for (const [url, bytes] of wastedBytesByUrl.entries()) {
```

Comment:
@patrickhulce almost missed this -- these lengths need to be transfer size, originally I had resource size.

I'm not confident in this code ...